code/hsec-sync/src/Security/Advisories/Sync/Git.hs

```diff
@@ -0,0 +1,163 @@
+data GitError = GitError
+  { gitCmd :: Cmd,
+    gitError :: GitErrorCase
+  }
+
+runGit :: [String] -> IO (Cmd, ExitCode, String, String)
+runGit args =
+  (\(status, stdout, stderr) -> (("git", args), status, stdout, stderr))
+    <$> readProcessWithExitCode "git" args ""
+
+data GitErrorCase
+  = -- | exit code, stdout and stderr
+    GitProcessError ExitCode String String
+  | -- | unable to parse this input as a datetime
+    GitTimeParseError String
+  deriving (Show)
+
+explainGitError :: GitError -> String
+explainGitError e =
+  unlines
+    [ "Called " <> show (fst $ gitCmd e) <> " with " <> show (snd $ gitCmd e),
+      case gitError e of
+        GitProcessError status stdout stderr ->
+          unlines
+            [ "git exited with status " <> show status,
+              ">>> standard output:",
+              stdout,
+              ">>> standard error:",
+              stderr
+            ]
+        GitTimeParseError s ->
+          "failed to parse time: " <> s
+    ]
+
+data Repository = Repository
+  { repositoryRoot :: FilePath,
+    repositoryUrl :: String,
+    repositoryBranch :: String
+  }
+
+data GitRepositoryStatus
+  = GitDirectoryMissing
+  | GitDirectoryEmpty
+  | GitDirectoryInitialized
+
+gitRepositoryStatus :: Repository -> IO GitRepositoryStatus
+gitRepositoryStatus repo = do
+  exists <- D.doesDirectoryExist $ repositoryRoot repo
+  if exists
+    then D.withCurrentDirectory (repositoryRoot repo) $ do
+      (_, checkStatus, checkStdout, _) <-
+        runGit ["rev-parse", "--is-inside-work-tree"]
+      let out = filter (not . null) $ lines checkStdout
+      case checkStatus of
+        ExitSuccess
+          | not (null out) && head out == "true" ->
+              return GitDirectoryInitialized
+        _ ->
+          return GitDirectoryEmpty
+    else return GitDirectoryMissing
+
+data GitRepositoryEnsuredStatus
+  = GitRepositoryCreated
+  | GitRepositoryExisting
+
+ensureGitRepositoryWithRemote ::
+  Repository ->
+  GitRepositoryStatus ->
+  IO (Either GitError GitRepositoryEnsuredStatus)
+ensureGitRepositoryWithRemote repo =
+  \case
+    GitDirectoryMissing ->
+      clone
+    GitDirectoryEmpty ->
+      clone
+    GitDirectoryInitialized ->
+      return $ Right GitRepositoryExisting
+  where
+    clone = do
+      (cmd, status, stdout, stderr) <-
+        runGit ["clone", "-b", repositoryBranch repo, repositoryUrl repo, repositoryRoot repo]
+      return $
+        if status /= ExitSuccess
+          then Left $ GitError cmd $ GitProcessError status stdout stderr
+          else Right GitRepositoryCreated
+
+updateGitRepository :: Repository -> IO (Either GitError ())
+updateGitRepository repo =
+  D.withCurrentDirectory (repositoryRoot repo) $ do
+    _ <- runGit ["remote", "add", "origin", repositoryUrl repo] -- can fail if it exists
+    (setUrlCmd, setUrlStatus, setUrlStdout, setUrlStderr) <-
+      runGit ["remote", "set-url", "origin", repositoryUrl repo]
+    if setUrlStatus /= ExitSuccess
```

Comment:
I believe that some, if not most, of these early returns can be replaced with `when`. (also you should be using `pure`, for reasons of forward-compatibility.)

Reply:
IIUC `when` does not shortcut/early return in `IO`.

Reply:
So, for context: they could be replaced by when guards and have the appropriate early return effect if there was a mechanism of exception handling that existed. Very much something I expect from my usual web application business in Haskell so much that I didn't realise I had come to expect the same structure everywhere else without explanation. (I will not move an inch on this matter of pure vs. return but you know I love you dearly nonetheless)

Reply:
You could just use ExceptT

Reply:
```haskell
bla :: IO ()
bla =  print =<< runExceptT do
  when True do 
    throwError "oh no" 
  liftIO $ print "hello"
```

I agree the nesting is not beautiful...